<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Introduction to Bioinformatics</title>

<meta name="description" content="Introduction to Bioinformatics">

<!-- Disabled for now - embeds HTML in author string -->
<!--  <meta name="author" content="Lesson 1 - Gestalt &amp; remote access" /> -->

<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="reveal.js/css/reveal.css">
  <link rel="stylesheet" href="src/theme/csmall.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) +
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->

<!-- Note escaped dollars -->
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
    });
  </script>

</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Introduction to Bioinformatics</h1>
<h3>Lesson 1 - Gestalt &amp; remote access</h3>
<p>
<h4></h4>
</p>
</section>


<section id="introductions" class="level1">
<h1>Introductions</h1>
<ul>
<li>Name</li>
<li>What you’re working on</li>
<li>Experience</li>
</ul>
</section>
<section id="what-youll-get-from-this-class" class="level1">
<h1>What you’ll get from this class</h1>
<ul>
<li>High level overview of how to <em>approach</em> bioinformatics</li>
<li>Basic bioinformatics technical skills</li>
<li>Enough momentum to initiate self-learning process</li>
</ul>
</section>
<section id="what-you-wont-get-from-this-class" class="level1">
<h1>What you won’t get from this class</h1>
<p>…but that might be useful:</p>
<ul>
<li>Higher level programming</li>
<li>Statistics</li>
<li>Other mathematics</li>
<li>Knowledge of every bioinformatics program/method ever created</li>
</ul>
<p>But hopefully you’ll have enough momentum to learn these on your own as needed.</p>
</section>
<section id="please-spend-some-time-outside-class" class="level1">
<h1>Please spend <em>some</em> time outside class</h1>
<ul>
<li>Time is the only way to learn these things</li>
<li>The waiting list is &gt; 150 long</li>
</ul>
</section>
<section id="overview-of-class" class="level1">
<h1>Overview of class</h1>
<ol type="1">
<li>Gestalt &amp; remote access</li>
<li>Unix shell - overview/intro</li>
<li>Unix shell - scripting &amp; advanced tools</li>
<li>Version control with git</li>
<li>Python - syntax &amp; types</li>
<li>Python - data structures</li>
<li>Python - writing a program</li>
</ol>
</section>
<section id="today" class="level1">
<h1>Today</h1>
<ul>
<li>1/2 overview</li>
<li>1/2 remote access &amp; practicum</li>
</ul>
</section>
<section id="why-computers" class="level1">
<h1>Why computers?</h1>
<p><br/></p>
<p><strong>DATA</strong></p>
</section>
<section id="biology-data" class="level1">
<h1>Biology data</h1>
<p>We can now get it</p>
<ul>
<li>Cheaper &amp; Faster</li>
</ul>
<p>=&gt;</p>
<ul>
<li>We have more of it</li>
</ul>
</section>
<section id="to-understand-biology-we-need-to-understand-data" class="level1">
<h1>To understand biology, we need to understand data</h1>
<p><br/></p>
<p><strong>ENTER BIOINFORMATICS</strong></p>
</section>
<section id="what-is-bioinformatics" class="level1">
<h1>What is bioinformatics?</h1>
</section>
<section id="from-the-book" class="level1">
<h1>From the book</h1>
<p>(edited)</p>
<blockquote>
<p>A jazz musician can walk into a nightclub, hear a familiar song, recognize the chord changes, and begin playing musical ideas over these chords. Likewise, a bioinformatician may receive a huge sequencing data set, and immediately start using a set of tools to learn about the data, and see what story the data tell.</p>
<p>… In both cases learning one’s tools is a necessary, but not sufficient step…</p>
</blockquote>
</section>
<section id="from-the-bird-charlie-parker" class="level1">
<h1>From the “Bird”, Charlie Parker</h1>
<blockquote>
<p>You’ve got to learn your instrument. Then, you practice, practice, practice. And then, when you finally get up there on the bandstand, forget all that and just wail.</p>
</blockquote>
<p>Also phrased as</p>
<blockquote>
<p>Learn the changes, then forget them.</p>
</blockquote>
</section>
<section id="my-summary" class="level1">
<h1>My summary</h1>
<section id="bioinformatics-is-more-than-the-sum-of-the-pieces" class="level2">
<h2>Bioinformatics is more than the sum of the pieces</h2>
<p>There is an art to putting the technical pieces together to make science happen.</p>
</section>
</section>
<section id="aspects-of-bioinformatics" class="level1">
<h1>Aspects of Bioinformatics</h1>
<p>Categories of things we typically do:</p>
<p>munging | exploration | conclusions</p>
</section>
<section id="data-munging" class="level1">
<h1>Data munging</h1>
<ul>
<li>Cleaning data</li>
<li>Transforming data</li>
</ul>
</section>
<section id="data-exploration" class="level1">
<h1>Data exploration</h1>
<ul>
<li>Visualize</li>
<li>Build intuition</li>
<li>Hypothesize</li>
</ul>
</section>
<section id="building-conclusions" class="level1">
<h1>Building conclusions</h1>
<p>In which we shore things up to tell the world…</p>
<ul>
<li>Statistical analysis</li>
<li>More visualization</li>
</ul>
</section>
<section id="so-this-is-what-we-do" class="level1">
<h1>So this is <em>what</em> we do</h1>
<p><br/></p>
<section id="but-how-should-we-do-it" class="level2">
<h2>But <em>how</em> should we do it?</h2>
</section>
</section>
<section id="specific-goals-for-research" class="level1">
<h1>Specific goals for research</h1>
<ul>
<li>Reproducibility</li>
<li>Robustness</li>
<li>Accuracy</li>
<li>Clarity</li>
<li>Iterability</li>
</ul>
</section>
<section id="things-we-can-do-towards-these-goals" class="level1">
<h1>Things we can do towards these goals</h1>
<p>immutability | documentation | automation</p>
</section>
<section id="treat-data-as-immutable" class="level1">
<h1>Treat data as immutable!</h1>
<p><strong>Never</strong> change data in place; data in, data out.</p>
<p>Aids: reproducibility, clarity &amp; iterability</p>
</section>
<section id="document-everything" class="level1">
<h1>Document <strong>everything</strong>!!!</h1>
<p>Hopefully you’re already doing this with your biology work.</p>
<p>Bioinformatics should be no different.</p>
<ul>
<li>Software versions</li>
<li>Settings used to run things</li>
<li>Reasoning behind the way you do things</li>
<li>How/when you got your data</li>
</ul>
<p>Aids: reproducibility, clarity</p>
</section>
<section id="the-best-documentation-is-automation" class="level1">
<h1>The best documentation is automation</h1>
<p>Write <em>scripts</em> for everything:</p>
<ul>
<li>Munging</li>
<li>Plotting</li>
<li>Statistics</li>
<li>Build scripts (tying it all together)</li>
</ul>
<p>Aids: reproducibility, robustness, accuracy, iterability</p>
</section>
<section id="guiding-principals-practices-towards-this-end" class="level1">
<h1>Guiding principals &amp; practices towards this end</h1>
<p>Unix philosophy | project organization</p>
</section>
<section id="the-unix-philosophy" class="level1">
<h1>The Unix Philosophy</h1>
<ul>
<li>Small, composable tools that do one thing right</li>
<li>Plumb tools together into pipelines/scripts</li>
<li>Embrace plain text/data (no excel/docx)</li>
</ul>
<p>Bioinformatics naturally embraces this (more so than most data science).</p>
</section>
<section id="project-layout" class="level1">
<h1>Project layout</h1>
<p>Bottom line: Being consistent aids in automation &amp; clarity:</p>
<ul>
<li>naming schemes</li>
<li>path schemes</li>
</ul>
<p><em>Note: I use a somewhat different organizational scheme than the book’s author. Find what works for you…</em></p>
</section>
<section id="project-layout-1" class="level1">
<h1>Project layout</h1>
<p>Here’s how I typically organize things:</p>
<pre class="text"><code>your-project
├── README.md
├── build.sh
├── data
│   ├── sequences.fasta
│   └── metadata.csv
├── scripts
│   ├── clean.py
│   └── plot.R
└── output
    ├── alignment.fasta
    ├── cleaned_metadata.csv
    ├── tree.newick
    └── tree_plot.png</code></pre>
</section>
<section id="practicum" class="level1">
<h1>Practicum!</h1>
<p>But first, <strong>Questions?</strong></p>
</section>
<section id="fire-up-nomachine" class="level1">
<h1>Fire up NoMachine</h1>
<p>Open no machine, and click the “New” button in the orange bar. Name your connection, so you’ll remember what it is (lynx/manx).</p>
<ol type="1">
<li>The protocol is <strong>NX</strong></li>
<li>The host is <strong>lynx</strong> or <strong>lynx.fhcrc.org</strong> or <strong>manx</strong> or <strong>manx.fhcrc.org</strong>, port <strong>4000</strong></li>
<li>The authentication is <strong>Password</strong></li>
<li><strong>Don’t use a proxy</strong></li>
<li>Name your connection so you can remember what it is.</li>
</ol>
</section>
<section id="open-a-terminal" class="level1">
<h1>Open a terminal</h1>
<ol type="1">
<li>Click the square button in the upper left.</li>
<li>Type in “Terminal”</li>
<li>Click on the Terminal icon (the one with <code>&gt;_</code> on it)</li>
</ol>
</section>
<section id="brief-overview-of-a-unix-command" class="level1">
<h1>Brief overview of a Unix command</h1>
<p><code>command &lt;flags&gt; &lt;operands&gt;</code></p>
<p>Example: <code>ls -a ~</code></p>
<ul>
<li><code>ls</code>: The command, for listing directory contents</li>
<li><code>-a</code>: A flag (short), which specifies that hidden files should be listed</li>
<li><code>~</code>: An operand, in this case a special symbol which points to your home directory</li>
</ul>
</section>
<section id="ssh-into-rhino" class="level1">
<h1>SSH into rhino</h1>
<ol type="1">
<li>Enter <code>ssh rhino</code> into the terminal</li>
<li>Enter your Hutch password when prompted</li>
<li>Note the machine to which you’re connected (e.g. <code>rhino02</code> v <code>rhino04</code>)</li>
</ol>
</section>
<section id="download-tmux.conf-fire-up-tmux" class="level1">
<h1>Download tmux.conf &amp; fire up tmux</h1>
<pre><code>wget https://raw.githubusercontent.com/fredhutchio/intro-bioinformatics/gh-pages/config/tmux.conf

ls

mv tmux.conf ~/.tmux.conf

ls
ls -a

tmux</code></pre>
</section>
<section id="quick-tmux-demo" class="level1">
<h1>Quick tmux demo</h1>
<ul>
<li><code>Ctrl-a</code> - your “Command key”</li>
<li><code>Ctrl-a ?</code> - Help</li>
<li><code>Ctrl-a d</code> - Detatch session</li>
</ul>
<p><br/></p>
<ul>
<li><code>Ctrl-a c</code> - New window</li>
<li><code>Ctrl-a &lt;Space&gt;</code> - Next window</li>
</ul>
</section>
<section id="create-a-project-directory" class="level1">
<h1>Create a project directory</h1>
<pre><code>mkdir ~/bioinfclass

cd ~/bioinfclass
pwd

mkdir data output scripts

touch README.md</code></pre>
</section>
<section id="dowload-data-with-wget" class="level1">
<h1>Dowload data with <code>wget</code></h1>
<pre><code>wget https://raw.githubusercontent.com/fredhutchio/intro-bioinformatics/gh-pages/data.tar</code></pre>
</section>
<section id="a-bit-about-tar" class="level1">
<h1>A bit about tar…</h1>
<p><a href="figures/tar.png" target="_blank"><img src="figures/tar.png" height="300"></a></p>
</section>
<section id="its-not-that-bad" class="level1">
<h1>It’s not <em>that</em> bad</h1>
<pre><code>tar -vxf data.tar</code></pre>
<p>This unpacks the data into the <code>data</code> dir for us.</p>
<p>Use <code>tar --help</code> when you need help.</p>
<p>(Also see <code>gunzip</code> and <code>gzip</code> for <code>.zip</code> files.)</p>
</section>
<section id="lets-look-at-what-weve-done" class="level1">
<h1>Let’s look at what we’ve done</h1>
<pre><code>tree</code></pre>
</section>
<section id="first-lets-explore-the-sequence-file" class="level1">
<h1>First let’s explore the sequence file</h1>
<pre><code>cat data/sfv.fasta</code></pre>
</section>
<section id="now-with-less" class="level1">
<h1>Now with <code>less</code></h1>
<pre><code>less data/sfv.fasta</code></pre>
</section>
<section id="now-lets-look-at-the-metadata" class="level1">
<h1>Now let’s look at the metadata</h1>
<pre><code>less data/sfv.csv</code></pre>
</section>
<section id="now-with-column--t" class="level1">
<h1>Now with <code>column -t</code></h1>
<pre><code>column -t -s , data/sfv.csv</code></pre>
<ul>
<li><code>-t</code> Output as tab separated</li>
<li><code>-s ,</code> Read the input data as comma separated</li>
</ul>
</section>
<section id="now-with-column--t-and" class="level1">
<h1>Now with <code>column -t</code> and <code>|</code></h1>
<pre><code>column -t -s , data/sfv.csv | less -S</code></pre>
<p><code>|</code> takes the output of one command and passes it as input to the next (covered in more detail next class). This is part of the Unix philosophy’s composability.</p>
<p>Use <code>q</code> to exit from <code>less</code>.</p>
</section>
<section id="ready-for-takeoff" class="level1">
<h1>Ready for takeoff…</h1>
<p>SciComp has set up a lovely “module” system that allows us to set up custom environments.</p>
<p>For this class we will be using the <code>intro-bio</code> module, loaded by executing <code>module load intro-bio</code>, which sets things up <em>just for the duration of that shell session</em>.</p>
<p>When you issue this command, you should see something like:</p>
<pre><code>matsen@rhino01:~$ module load intro-bio
Using already loaded python
matsen@rhino01:~$</code></pre>
<p>If you get an error message instead, let us know, because you can’t proceed without this module loaded.</p>
</section>
<section id="hello-world" class="level1">
<h1>Hello world!</h1>
<section id="make-an-alignment-and-then-a-tree-from-the-data" class="level2">
<h2>Make an alignment, and then a tree from the data</h2>
<pre><code>muscle -maxiters 2 -in data/sfv.fasta -out output/alignment.fasta

FastTree -nt output/alignment.fasta &gt; output/tree.nw

nw_display output/tree.nw | less</code></pre>
</section>
</section>
<section id="resources" class="level1">
<h1>Resources</h1>
<ul>
<li>These slides: <a href="http://fredhutchio.github.io/intro-bioinformatics">http://fredhutchio.github.io/intro-bioinformatics</a></li>
<li>Unix command reference: <a href="https://ubuntudanmark.dk/filer/fwunixref.pdf">https://ubuntudanmark.dk/filer/fwunixref.pdf</a></li>
<li>Tmux tutorial: <a href="http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex">http://www.fredhutch.io/articles/2014/04/27/terminal-multiplex</a></li>
</ul>
</section>
</div>

<script src="reveal.js/lib/js/head.min.js"></script>
<script src="reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: false,


    rollingLinks: false,
  
  // available themes are in /css/theme
      theme: Reveal.getQueryHash().theme || 'csmall',
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'fade',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'reveal.js/plugin/markdown/showdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'}
    // { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });
Reveal.addEventListener( 'slidechanged', function( event ) {
    MathJax.Hub.Rerender();
} );
</script>

</body>
</html>
